frodo.js

```diff
@@ -1,91 +1,85 @@
 /**
```

Comment:
SCRIPT-You did a great job handling all the possible error states in the UI. It enhances the user experience significantly.